File: ui/backend/schemas/import_geometry.py
"""Pydantic schemas for the M5 STL case-import surface.

Mirrors the dataclass returns from ``ui.backend.services.geometry_ingest`` in
a JSON-serializable form. The route returns these to the frontend so
``ImportPage.tsx`` can render the ingest report card.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


UnitGuess = Literal["m", "mm", "in", "unknown"]


class PatchInfoPayload(BaseModel):
    name: str
    face_count: int


class IngestReportPayload(BaseModel):
    is_watertight: bool
    bbox_min: tuple[float, float, float]
    bbox_max: tuple[float, float, float]
    bbox_extent: tuple[float, float, float]
    unit_guess: UnitGuess
    solid_count: int
    face_count: int
    is_single_shell: bool
    patches: list[PatchInfoPayload]
    all_default_faces: bool
    warnings: list[str]
    errors: list[str]


class ImportSTLResponse(BaseModel):
    case_id: str
    ingest_report: IngestReportPayload
    edit_url: str


class ImportRejection(BaseModel):
    """Returned as 4xx body when ingest rejects."""

    reason: str
    failing_check: str
    ingest_report: IngestReportPayload | None = None

File: ui/frontend/src/types/import_geometry.ts
// Mirrors ui/backend/schemas/import_geometry.py — kept thin so the wire
// types stay one source-of-truth on the backend.

export type UnitGuess = "m" | "mm" | "in" | "unknown";

export interface PatchInfo {
  name: string;
  face_count: number;
}

export interface IngestReport {
  is_watertight: boolean;
  bbox_min: [number, number, number];
  bbox_max: [number, number, number];
  bbox_extent: [number, number, number];
  unit_guess: UnitGuess;
  solid_count: number;
  face_count: number;
  is_single_shell: boolean;
  patches: PatchInfo[];
  all_default_faces: boolean;
  warnings: string[];
  errors: string[];
}

export interface ImportSTLResponse {
  case_id: string;
  ingest_report: IngestReport;
  edit_url: string;
}

export interface ImportRejectionDetail {
  reason: string;
  failing_check: "stl_parse" | "watertight" | "size_limit" | "unknown" | string;
  ingest_report?: IngestReport | null;
}

// M-PANELS Step 10 · demo-fixtures catalogue. Mirrors the
// DemoFixture pydantic model in ui/backend/routes/demo_fixtures.py.
export interface DemoFixture {
  name: string;
  filename: string;
  title: string;
  description: string;
  size_bytes: number;
  // Phase-1A (DEC-V61-097): closed cavities (ldc_box) walk Steps 1→5
  // end-to-end; external-flow geometries (cylinder, naca0012) stop
  // at Step 2 because the gmsh interior mesh isn't a flow domain.
  full_demo_capable: boolean;
  capability_note: string;
}
